### ISSUE TITLE
Upstream status

### ISSUE THREAD

**Issue by Fantu** [veeam/blksnap#2]
@SergeiShtepa Hi, thanks for your works on blksnap kernel module, I think will be very useful for improve backup on linux.
I'm curious about the status of its possible upstream integration, from https://github.com/veeam/blksnap/commit/85bad8f010d08fb0d08fc2c33e5ca3630dfe2a82 seems you prepared a patches serie for upstream (for send a mail with it) but I not found it posted searching on lkml.org and google. I saw also other series of months ago on other branches but I not found them on upstream mailing list and found only older your works of 1 year ago or more.
Thanks for any reply and sorry for my bad english.

**Comment by SergeiShtepa** [comment]
Thank you for creating the issue. So it's important to someone.

Yes, I haven't tried to offer blksnap in the upstream yet.
The problem is that I can't find anyone who could do a code review. And I don 't have enough experience with upstream.
In addition, the module is not a regular driver for some device. I suggest adding a filter to the block layer. Mantainers may not like it.
It would be great if I could attract other developers of backup tools to work on the module. But I think that while the module is not in the upstream, no one will want to invest their time in this project.
Therefore, I doubt the success of this venture.

But I intend to try to do it soon. It remains to double-check something.
I hope that someone will give good advice.
I will add a link to the patch discussion in upstream to this issue.

Any feedback is welcome.

**Comment by Fantu** [comment]
Thanks for reply, I did/do small contributions on some open source projects but I never did on linux kernel, if I remember good I gave only an help in testing one xen part preparing to upstream many years ago.
To know if can be acceptable your work and receive review of expert kernel developers (who can give you their opinions on it and their advices to improve where needed) post the patches is really important, from a fast look I found this doc that explain how to do it: https://www.kernel.org/doc/html/latest/process/submitting-patches.html (probably you already saw it)
and from a fast look to your latest patches serie ( https://github.com/veeam/blksnap/commit/85bad8f010d08fb0d08fc2c33e5ca3630dfe2a82 ) seems you already did a good work on major requirements.
I found this project in a search done after I found one blk-snap patches serie you posted 2 year ago so I think posting patches is also a way let know it to any other interested developers and users (as well as receiving upstream developers review)

**Comment by SergeiShtepa** [comment]
> seems you already did a good work on major requirements. 
Thanks. Indeed, it is the good quality of the code that is an important component of success.
> a way let know it to any other interested developers and users
Of course, that's why I'll be offering a patch.

**Comment by andymend** [comment]
I agree that your code represents a potentially important improvement for backup of linux systems, especially as new file systems proliferate. Blksnap is a distinct improvement over earlier such attempts like dattobd and elastio-snap.  I have been lightly testing  blksnap on a computer I routinely use where there are multiple file systems being used. Yours is a worthwhile project that will hopefully get merged into the mainstream kernel at some point.

**Comment by SergeiShtepa** [comment]
Thank you @andymend . 
Thank you all for your support.
I just sent a patch.
If it passes the robot check, we will see it in the [list](https://patchwork.kernel.org/project/linux-block/list/) soon.

**Comment by SergeiShtepa** [comment]
https://patchwork.kernel.org/project/linux-block/list/?series=649931

**Comment by Fantu** [comment]
for what I see [here](https://lore.kernel.org/linux-block/1655135593-1900-1-git-send-email-sergei.shtepa@veeam.com/) 2 kernel developers reviewed and replied and there is also 2 replied from an automatic test bot (reporting some warnings), for now, no one has objected or criticized the implementation method but only recommended improvements for a v2.

from a quick search I did not find other proposals for implementing "temporary" snapshots for better backup integrity and that supports many filesystems

unfortunately I have not yet had time to try blksnap but from what I have seen from experience this functionality integrated into the kernel is very useful and as I have already seen several backup solutions have made their own modules for similar things but no one who has proposed its implementation in the kernel

I have used several backup solutions over the years, most with file level backup using rsync, however even in such cases I think it would be useful because either the services stop during the whole backup or you risk not having it integrated and the larger the part to back up and the more time it takes, the greater the risk of integrity being compromised.
For example, if you have to back up a server of hundreds of gb or over 1 tb (in most cases via the network the times increase), it takes hours and during the operations to numerous files in various interconnected parts and rsync buckups of the parts before such them, others during and others after would be problematic for its integrity on restore while with a snapshot and rsync backup on it you would have a backup with minimal integrity risk (reduced to the fraction of snapshot creation time FWIK) and restore you would know that you have a fairly healthy situation at the time of the snapshot and not a high risk of integrity problems for the time it took to make the backup. And without such a snapshot, the risks can only be mitigated by stopping the services (as mentioned) or doing it outside working hours.

Another thing I tried at works in latest years is to use btrfs with integrated snapshot for backups (I also tried zfs years ago) but I use only in some cases because for example on virtualization systems, the COW it has a significant impact on performance and there are also some space issues keeping snapshots on the same production filesystem. this could be a solution to make a backup with minimum downtime stopping services to take the snapshot (with blksnap) and maximum integrity probability, or at least making the snapshot to active services like what a COW filesystem like btrfs and zfs can do but without having to use such filesystems.

I hope these examples based on my experience give an idea of how useful a system like blksnap is
When there is a calm period where I am not too busy I will start testing it
in the meantime thanks again for your work SergeiShtepa

sorry for my bad english

**Comment by SergeiShtepa** [comment]
Thanks @Fantu.

I expected a very calm reaction to the patch.
I plan to:
* correct a flaws, indicated by the reviewers
* move some of the logic into the kernel module, simplifying the main kernel code even more
* fix bugs (QA found something).

So, the next patch will be better.

Alas, the performance of the blksnap module should not be better than DM snapshots (LVM2).
When handling a write request, it is required to read a chunk of data, and then write this chunk of data to a new location. There are no magic here. When writing, the load on the drive increases threefold.

Theoretically, snapshots on Btrfs should be better. But for BTRFS, we cannot read the entire block device. We have to synchronize the source and target file systems. The Veeam Agent for Linux implements btrfs backup support by its native means. The performance was not the best.

The presence of a change tracker and the ease of allocating space for the difference storage are the main advantages of blksnap, compared to "classic" snapshots.

**Comment by Fantu** [comment]
thanks for reply and your works
about btrfs I know that is better where there is no "limit" or performance issue (for example I use on some backup storage and fileservers), but for example in "low cost" virtualization servers there was  significant performance issue, disabling COW partially solves the issue because using snapshot will require it and ignore the previous options to disable cow, so still using ext4 and do temp. snapshot only on backup time, avoid the downtime or having it shortly, having high integrity probability and lower performance only when backup with snapshot is running still seems to me a good idea.
however I can only be helpful enough when I start using it

**Comment by Fantu** [comment]
hi, is there any news about v2 patch serie for upstream?
I suppose anyway it would not be reviewed in time for the kernel 6.0 :(

**Comment by SergeiShtepa** [comment]
Hi. That's what I really want to do.
But I had to switch to more urgent tasks. I hope to be able to reduce my task list soon.

**Comment by SergeiShtepa** [comment]
Hi!

[This](https://github.com/veeam/blksnap/commit/6ca28c5370424c7b01f5ec39db4e8b9ff873f3f4) looks like a workable version of the patch for [linux kernel 6.1-rc1](https://git.kernel.org/torvalds/t/linux-6.1-rc1.tar.gz) . 
It contains corrections in accordance with the comments that were made by the maintainers to the first version of the patch.
I plan to test this for two or three days.
I invite everyone to join the testing. 

---
Feedback is welcome.

**Comment by SergeiShtepa** [comment]
This [patch](https://github.com/veeam/blksnap/tree/master/patches/lk6.1-rc3) has been sent.

**Comment by Fantu** [comment]
good, I not checked it in detail but I noted a small mistake on the version of the patch serie posted, v1 instead of v2, but if you have sent it by now there is nothing else to do (or they would receive it double), I saw there is at least a list of changes from v1 in the cover so it should be "ok" anyway

**Comment by SergeiShtepa** [comment]
Link to [patchwork](https://patchwork.kernel.org/project/linux-block/list/?series=691286).

**Comment by Fantu** [comment]
@SergeiShtepa with a search I found this: https://lwn.net/Articles/914852/ and seems that wrote about a lack of documentation to be accepted
in addition to the documentation in code that probably need improvements I suppose they meaning to add another "generic" in Documentation/, I suppose in Documentation/block creating Documentation/block/blksnap.rst and adding in the list of Documentation/block/index.rst
I don't saw reply to lack of documentation in the replies of the patch serie posted, you received other reply directly and not in mailing list?

**Comment by SergeiShtepa** [comment]
Thanks.
I don't know anything about it.
I received one email about an extra space and 3 emails from the robot.
I'll try to add a couple of comments to the article.

**Comment by Fantu** [comment]
@SergeiShtepa I saw your replies on lwn, unfortunately even as you explained you want to avoid additional time being wasted reviewing documentation before it is accepted upstream I suppose without more documentation some people won't start reviewing it. For now on mailing list of patch posted I didn't see new replies :(
I did a fast check with kernel-doc script of the kernel-doc comments actually present and spotted some warning:
one I suppose solved with https://github.com/Fantu/linux-blksnap/commit/4f7c2365bcd5acc27a7bd546dc20b927fbf1b576
others spotted:
```
drivers/block/blksnap/tracker.h:47: warning: Function parameter or member 'flt' not described in 'tracker'
drivers/block/blksnap/tracker.h:47: warning: Function parameter or member 'submit_lock' not described in 'tracker'

drivers/block/blksnap/diff_storage.c:21: warning: Function parameter or member 'link' not described in 'storage_bdev'
drivers/block/blksnap/diff_storage.c:21: warning: Function parameter or member 'dev_id' not described in 'storage_bdev'
drivers/block/blksnap/diff_storage.c:21: warning: Function parameter or member 'bdev' not described in 'storage_bdev'

drivers/block/blksnap/diff_storage.c:33: warning: Function parameter or member 'link' not described in 'storage_block'
drivers/block/blksnap/diff_storage.c:33: warning: Function parameter or member 'bdev' not described in 'storage_block'
drivers/block/blksnap/diff_storage.c:33: warning: Function parameter or member 'sector' not described in 'storage_block'
drivers/block/blksnap/diff_storage.c:33: warning: Function parameter or member 'count' not described in 'storage_block'
drivers/block/blksnap/diff_storage.c:33: warning: Function parameter or member 'used' not described in 'storage_block'
```

**Comment by SergeiShtepa** [comment]
To read the entire article [https://lwn.net/Articles/914031/](https://lwn.net/Articles/914031/) and comments before December 1, you need to be a registered user. After December 1st, access will be unlimited. I tried to answer all the questions.
I plan to work on the documentation and try to write an article for LWN. Maybe it will help.
> spotted some warning:

Thanks, I'll have to check it out too for next patch. 

Perhaps the next v2 patch with documentation and accompanied by an interesting article will attract more attention.

**Comment by Fantu** [comment]
@SergeiShtepa thanks for replies and your work
in practice v2 would be the last one posted and v3 the next even if it was written v1, I wouldn't want that perhaps having written v1 had led some people to suppose in a resend without changes, for example it seems strange to me that Christoph Hellwig who had participated a lot in the the first did not give any answer to the second (probably can be good add it in cc of the next)
looking also on other sites the article with write about "the lack of doc." seems to have left its mark :( 
https://www.reddit.com/r/linux/comments/yvd14a/blockdevice_snapshots_with_blksnap_lwnnet_good/
probably a v3 with more documentation and even just with the bot's warning fixes and Randy's reporting will push sites to update about it and attract more people to review

**Comment by Fantu** [comment]
@SergeiShtepa in the patch serie is missed the add of entry in [MAINTAINERS](https://github.com/Fantu/linux-blksnap/blob/upstream-propose-v2/MAINTAINERS) file

**Comment by SergeiShtepa** [comment]
> @SergeiShtepa in the patch serie is missed the add of entry in [MAINTAINERS](https://github.com/Fantu/linux-blksnap/blob/upstream-propose-v2/MAINTAINERS) file

Maybe... 
I'll add this in the next patch.

**Comment by SergeiShtepa** [comment]
Add docs for /linux/Documentation/block/ : [blkfilter](https://github.com/veeam/blksnap/blob/master/doc/blkfilter.rst) and [blksnap](https://github.com/veeam/blksnap/blob/master/doc/blksnap.rst) .
I really hope that I managed to find the right words.

Any feedback is welcome.

P.S.: Since yesterday [article](https://lwn.net/Articles/914031/) is available without registration.

**Comment by Fantu** [comment]
@SergeiShtepa thanks for the creation of documentation to include in the patch for upstream
my english isn't great but i read them, blkfilter it seems to be ok, about blksnap there is a non-english sentence in "Snapshot overflow resistance" and the part of defects of the alternatives and merits of blksnap leaves me doubtful, I'm afraid that someone who will review or want to watch/try blksnap will get lost in counterproductive discussions.

about btrfs "Obviously cannot be applied to other file systems.", I suppose you mean the fact that the main advantages of transferring to another local or remote destination are lost from or to a different filesystem, and it is, but written like this is wrong, in practice it is possible to use it even if most of the advantages are lost. You can use it both for "temporary" snapshots during which you make a backup (for example with rsync) to another destination; and from a source without btrfs, you make backups with another (for example rsync) and on the destination you use the btrfs snapshots to manage the "history" with some advantages (I've been using it on some systems for years).

the rest of the disadvantages I saw in practice during a long testing some years ago thinking about using btrfs instead of ext4 on several virtualization systems to manage better and faster backups but the performance impact was huge (partly due to the use of hdd) and even if only the last corresponding snapshot was kept for send/receive with the destination there were disadvantages, reduced in part with a single destination and daily backup but considerable with multiple destinations (in my case 2) in "rotation" and having to keep a week or even more the snapshots.
I've seen what the arrival of the btrfs fs to full space caused by snapshot, the damage it causes especially with the virtual machine (which do not have direct access to the fs and "amplify" the problem) and also the fact that it is not "automatically managed/avoided" and the workaround of scripting to see if it's about to fill up and deleting the snapshot(s) caused most of the "PROs" to decay when that happens, which is why we didn't use it in production for that use case

however without listing all the possibilities I think it is wrong to write things like "the blksnap module is the best way to create snapshots for backup tools", it is in good part of cases but not all.
I'm not good at writing and I don't quite know how to describe such parts properly, objectively and correctly but I think that the documentation it should be done that way people should just understand **the need for this method, its importance for a good chunk of cases and the need to get it good and integrated upstream**

sorry if I didn't explain myself well and for my bad english

**Comment by SergeiShtepa** [comment]
@Fantu - Thanks for the feedback!

I will try to improve those sections that you have noticed.
After the weekend, I will review the documentation again.
Maybe my colleagues from the technical documentation department will help me.

My English is also not as good as I would like. So I'm trying to do what I'm capable of. :)
Creating documentation is hard work, and now it looks really important.

I can suggest you try to make the document better. Why not try?
Even if I don't accept your merge request, it will allow me to take a fresh look at some sections.

**Comment by Fantu** [comment]
@SergeiShtepa saw the changes, now is better, you missed to fix non-english sentence in "Snapshot overflow resistance"

**Comment by SergeiShtepa** [comment]
Prepared a patch [blksnap_lk6.1-rc8](https://github.com/veeam/blksnap/commit/6fb2b8882e5e52eaec42e08b2b922fbcefd5ad4b).
There are problems with kernel-docs. I'll fix it.
Everyone can try to build a patch, launch it, criticize it, suggest changes.
I will test the patch on different architectures and in different test labs.

**Comment by Fantu** [comment]
@SergeiShtepa thanks for your works
before I posted a comment linking doc files in the article: https://lwn.net/Articles/917121/
about documentation of next patch serie I did fast test (make htmldocs), link is missed, tried fast to add it and generate again: https://github.com/Fantu/linux-blksnap/commit/0ee9d76019867bdde70e513e10e57398f75991f9
works but there is a small strange thing in blksnap line in block index that I don't understand how to fix (https://snipboard.io/OJgr2s.jpg)

in maintainers file there is a small typo, I suppose will be good have also a git repo (like linux-blksnap to add in https://github.com/veeam) and add in it for make easier keep patch serie for user/dev use, test, rebase etc; for report specific issue, propose PR, after you (as blksnap maintainer) can review/add/improve them before post to kernel mailing list etc

**Comment by SergeiShtepa** [comment]
`make htmldocs` - I checked it out. it looks good.
The module in the patch cannot build. I will check how it work on my lab and will update patch.

**Comment by SergeiShtepa** [comment]
linux kenrel branch [blksnap_lk6.1-rc8](https://github.com/SergeiShtepa/linux/tree/blksnap_lk6.1-rc8).

**Comment by SergeiShtepa** [comment]
> FWIK the path submitted for upstream should before be tested with build and use trying with both module and builtin, in addition for a better quality it would be better to test also on different architectures and different operating systems (for different build software versions etc)

I'm just doing testing. With tests and as part of the VAL product. That's right now I'm compile kernel for ppc64el. I also involve QA in this task. They have their own labs and autotests. But I prefer Debian/Ubuntu for kernel testing.

> I don't know if it is possible to run entire kernel builds and automatic boot and use tests in basic and free CI testing of github, circleci and similar as they are very expensive, it will probably have to be configured on dedicated systems, I have never tried it but perhaps the [kernelci](https://github.com/kernelci) could be worth trying, especially to set up a complete CI when it will be accepted upstream

I have no experience in this field.

> another thing I think good to do will be prepare a better module build tests (for kernel >= 5.10 and <upstream version that will include blksnap) 

it is possible to use the [SUSE service](https://build.opensuse.org/project) to check compilability.

**Comment by SergeiShtepa** [comment]
Branch [blksnap_lk6.1-rc8_v2](https://github.com/SergeiShtepa/linux/tree/blksnap_lk6.1-rc8_v2) with patch for linux kernel.
Prepared [email with patch](https://github.com/veeam/blksnap/tree/master/patches/lk6.1-rc8-v2).
Regression testing is scheduled at night. If everything goes well, then tomorrow it will be possible to send a patch.

**Comment by Fantu** [comment]
good, I did a fast check of latest patch and with "make htmldocs" remained these 2 warning:
```
./include/uapi/linux/blksnap.h:243: warning: This comment starts with '/**', but isn't a kernel-doc comment. Refer Documentation/doc-guide/kernel-doc.rst
/home/fabio/repository/linux-blksnap/Documentation/block/blksnap.rst:139: WARNING: duplicate label block/blksnap:change tracker, other instance in /home/fabio/repository/linux-blksnap/Documentation/block/blksnap.rst
```
I don't know what of the 2 "Change tracker" title change and with what, also a minimal change to don't have same label (for link) I suppose is enough

relooking about blkfilter doc probably will be good do some addition, for example mention that is done as "single filter" (as requested by a block maintaner) and used by blksnap (and link to it doc)
and probably include also kernel-doc of the functions?
did a fast try (https://github.com/Fantu/linux-blksnap/commit/798b0f2c42dae8b5f822294245e24e89bcce1e58) but text need to be improved and link have a warning that need to be fixed:
`/home/fabio/repository/linux-blksnap/Documentation/block/blkfilter.rst:26: WARNING: undefined label: ../blksnap.rst (if the link has no caption the label must precede a section header)`

Spotted also 2 descriptions to be updated: https://github.com/SergeiShtepa/linux/commit/b4363f5891bf674be4a19ea1099375f9240e7cc1#r92333996 https://github.com/SergeiShtepa/linux/commit/b4363f5891bf674be4a19ea1099375f9240e7cc1#r92334083

**Comment by SergeiShtepa** [comment]
Thanks a lot for the feedback.

I realized that the documentation needs improvement. I'm working on it.

**Comment by SergeiShtepa** [comment]
@Fantu .
I have modified the documentation. There are many changes. [commit in my linux](https://github.com/SergeiShtepa/linux/commit/63bac1d2bf96c7fb5e1a652f7ba0084bd71a9f0f).
But now the html version of the document looks harmonious.
I haven't moved the changes to the master for the module yet.

**Comment by Fantu** [comment]
@SergeiShtepa good, thanks
I tested (make htmldocs) and now there is no warning related to blksnap and blkfilter, I spotted a typo (comment in the commit), checked blkfilter and blksnap html output generated and seems good, for me further expansions could also be made later but it will be necessary to see the opinion of the more expert, tried to ask to the article author: https://lwn.net/Articles/917482/ I'm curious to know his opinion, and/or others
About build kernel and test it (with the new version of the patch) I not had time for now, I hope to have time tomorrow.

about the change to Documentation/block/capability.rst seems the output (html) generated from your patch is already near the same of the one actual (https://docs.kernel.org/block/capability.html)
I suppose there is something related to kernel-doc that I don't know that already make select "DOC: genhd capability flags" part in include/linux/blkdev.h.
Is there a fix/improvement anyway I don't see from a fast look? if yes can be posted directly to the mailing list in a separate patch (separate from blksnap serie I mean)

**Comment by SergeiShtepa** [comment]
local [patch](https://github.com/veeam/blksnap/tree/master/patches/lk6.1-rc8-v5).
[branch](https://github.com/SergeiShtepa/linux/tree/blksnap_lk6.1-rc8_v5) for linux kernel.

**Comment by Fantu** [comment]
good, one people already replied about documentation part and there is still warnings/errors spotted by kernel test robot
I think we should do more for make reachable/visible the project, small and fast things as wrote here: https://github.com/veeam/blksnap/issues/21#issue-1465200362
I also did some small improvements to readme: https://github.com/veeam/blksnap/pull/28
But many other improvements and addition are needed to make both developers and users easy and fast start using/testing/contributing

**Comment by Fantu** [comment]
@SergeiShtepa I see a very old version of your patch posted: https://lkml.org/lkml/2020/10/21/122
and based on these: https://lkml.org/lkml/2020/10/21/128 https://lkml.org/lkml/2020/10/21/849
in https://lore.kernel.org/lkml/20221209142331.26395-3-sergei.shtepa@veeam.com/ is missed EXPORT_SYMBOL() -> EXPORT_SYMBOL_GPL()

**Comment by SergeiShtepa** [comment]
> is missed EXPORT_SYMBOL() -> EXPORT_SYMBOL_GPL()

I think it's not a matter of principle. Any non-GPL module can copy the body of the attachment function and also use the filter.

**Comment by Fantu** [comment]
@SergeiShtepa thanks for reply, sorry, I did not know about that

I saw also this about patch versioning: https://lkml.org/lkml/2020/10/24/108
version should always be incremented and include the changelog of all version posted
seeing the old patches I probably found why unfortunately many people no longer review the blksnap patches :(

however in the last few patches you made many progress to follow the guidelines, make the fixes and improvements requested by the maintainers, keep making better patches and I'm sure the revisions will increase and finally you will be able to integrate blksnap upstream :)

**Comment by Fantu** [comment]
@SergeiShtepa I re-read the cover of v2 and changelog part don't seems look very good, to mitigate the "error of v1" and clarify the changes perhaps for the next version I think can be changed with something like:

> The initial "without version" was posted at 13 June 2022. ...
> 
> Changes in v1:
> - Forgotten "static" declarations have been added.
> ...

> The v1 version was posted at 2 November 2022. ...
> 
> Changes in v2:

> The v2 version was posted at 9 Dicember 2022. ...
> 
> Changes in v3:

I also replaced "suggested" with "posted" but if you think suggested is better keep it

**Comment by Fantu** [comment]
I saw the replies of Christoph Hellwig, he gave another time a very good contribute.
I suppose that for next version should be more that only mention in the cover but something at least in the patches where he gave a significant contribute: https://docs.kernel.org/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by

**Comment by SergeiShtepa** [comment]
> I saw the replies of Christoph Hellwig, he gave another time a very good contribute.

Yep. It makes sense to continue working further. :)

> I suppose that for next version should be more that only mention in the cover but something at least in the patches where he gave a significant contribute: https://docs.kernel.org/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by

Yes. I will take this into account when preparing v3.
In any case, I will publish the patch in blksnap/patches/ in advance for verification.

**Comment by SergeiShtepa** [comment]
Hi.
There were a lot of changes in the master branch. Bug fixes from the VAL-6.0 branch and changes related with preparing v3 patch. So, something might not work :)
The module interface has changed, so the old binary code may not work.
Interface changes are not finished yet, the most interesting is yet to come.

**Comment by Fantu** [comment]
@SergeiShtepa thanks for your work, I suggest in the case like this to keep incomplete/not working work in a separate branch (but pushed) and not merging in master already or user will try blksnap first time and see the "bad state" can't know is a temp. work in progress
can be also useful open a PR (and mark draft if incomplete) for users/contributors want review a work in progress

**Comment by SergeiShtepa** [comment]
Thanks for the comment.
I use a workflow out of habit, where the master branch is an unstable development, and stable versions in separate branches, such as VAL-6.0. Perhaps this does not correspond to the generally accepted workflow. I'll think about it twice next time.
On the one hand, it is possible now to make a branch from the [commit](https://github.com/veeam/blksnap/commit/c25a730564420a07b48244c0b62c12c824cc6348) and rename it to master, and rename the current one to devel.
On the other hand, a stable master branch does not meet the goals of the repository. The main goal is to become part of the Linux kernel. Therefore, it makes no sense to keep a stable master branch.
So, I'm not sure which choice would be better.

I'm think to fix the first version 1.0.0. when module will be in upstream. The increment will be do only for after that. Perhaps by that time a standalone version will no longer be needed. Although the repository will of course remain to support the user space code. I think we will need a version at that stage of the project's life.

**Comment by Fantu** [comment]
@SergeiShtepa you are right that master branch is normally development/unstable but for incomplete work in progress I think can be better do in a branch itself until a more complete state (and open a PR for track comment/review etc..), this can be useful if users/developer will approcing to blksnap, instead there is a more high risk that consider blksnap project "too bad and escape"

I link an example: https://github.com/linuxmint/muffin/pull/601 was a rebase on newer mutter after near ten years from fork, was a more big task and was a work in progress for 2 major cinnamon versions (in the meantime an intermediate version on master continued to be developed as, especially at the beginning, there were many shortcomings and problems and it could only be used for specific development and tests)
A similar thing I suppose can be done with blksnap major changes, at least until "complete", I don't mean bug free, is ok merging in master even if still have many bugs but not very if is incomplete and definitely not working. separate branch and open PR I think can be done also for only few days needed to complete but you start to push the work done for now

maybe for this, leave it as it is, unfortunately the participation in the project is still almost nil (outside the ML) apart from a very small contribution from me :(

have also stable branches are good (you already keep one for example) then it would be good to have a versioning I was already thinking to open a specific "feature request" to talk about it, I'll do when I'll have more time

sorry for my bad english and if I not explained good

**Comment by SergeiShtepa** [comment]
ok. Thanks for the feedback.
I'll leave it as it is for now. In the future, I will not be lazy and will make feature-branches.

**Comment by Fantu** [comment]
@SergeiShtepa hi, is there any news about next version for upstream?
even if not all changes was done and patches rebased if you want already push like what was for example https://github.com/SergeiShtepa/linux/commits/blksnap_lk6.1-rc8
in the meantime I tried to prepare automatic tests to quickly test the build of the next versions, even if unfortunately I had difficulties with the other archs on amd64 both builtin and module are working and I also found the tests for ppc archs
v2 on ppc fails but seems for same errors already reported by "kernel test robot" https://github.com/Fantu/linux-blksnap/actions/runs/3824817943

**Comment by SergeiShtepa** [comment]
>  is there any news about next version for upstream?

Working with this [comment](https://lore.kernel.org/linux-block/Y5rwa6m3yqo40vz1@infradead.org/).
There is a lot of work to do. Interface changes change the logic of object ownership. I hope it will be better :-).

You don't even have to try to build standalone modules on any architectures other than amd64.
But the patch should work on arm64el and ppc64el.  I tested the patch with a fresh kernel on POWERPC. I haven't tried big endian yet.
```
./usr/include/linux/blksnap.h:508:15: error: expected declaration specifiers or '...' before 'sizeof'
  508 | static_assert(sizeof(struct blk_snap_snapshot_event) == 4096,
```
I don't understand yet what kind of problem static_assert creates.  I'll probably have to cut out these checks.

**Comment by Fantu** [comment]
@SergeiShtepa seems that static_assert should be changed in a way that compile in any case, from a fast search I found this probably useful:
https://stackoverflow.com/questions/3385515/static-assert-in-c/54993033#54993033
I don't have time to check in detail now, I hope can be useful, rr it would be to ask Christoph Hellwig if you can't find a solution

for all ppc build tests should be enough apply https://github.com/Fantu/linux-blksnap/commit/336b28ecf30e880a45c4b92eccb31bd697722338 after solve the static_assert error or I'll test after you push the new patch to git

**Comment by SergeiShtepa** [comment]
Work on patch v3 is underway.  To do this, I created a (branch)[https://github.com/sergeishtepa/linux/tree/blksnap-lk6.2-rc2].
The code has never been run yet. Perhaps it can be compiled.
The main change is the interface.

**Comment by Fantu** [comment]
build on amd64 with config similar to debian kernel packages was successfull both blksnap builtin and module...but after looking logs I saw don't built with blksnap because of changed config, my mistake of don't do more checks before so probably I'll post the result tomorrow

instead a "full config" build as needed prerequisite for kernel submit on ppc tests failed:
```
./usr/include/linux/blksnap.h:87:32: error: C++ style comments are not allowed in ISO C90
   87 |         __u64 device_capacity; // ? see BLKGETSIZE64
      |                                ^
./usr/include/linux/blksnap.h:87:32: note: (this will be reported only once per input file)
./usr/include/linux/blksnap.h:337:15: error: expected declaration specifiers or '...' before 'sizeof'
  337 | static_assert(sizeof(struct blksnap_snapshot_event) == 4096,
      |               ^~~~~~
./usr/include/linux/blksnap.h:338:9: error: expected declaration specifiers or '...' before string constant
  338 |         "The size struct blksnap_snapshot_event should be equal to the size of the page.");
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[4]: *** [/linux/usr/include/Makefile:91: usr/include/linux/blksnap.hdrtest] Error 1 
```

**Comment by Fantu** [comment]
@SergeiShtepa I did another fast search related to header error... in https://github.com/SergeiShtepa/linux/blob/master/usr/include/Makefile:
`# Unlike the kernel space, exported headers are written in standard C.`

So I tried a fast test with https://github.com/Fantu/linux-blksnap/commit/895a41fcdc62466a79b488e3e75e21f4c44c1869 and ppc tests "full" continued, I'll check result tomorrow

After I found also another thing in https://github.com/SergeiShtepa/linux/blob/master/include/linux/build_bug.h
```
/**
 * static_assert - check integer constant expression at build time
 *
 * static_assert() is a wrapper for the C11 _Static_assert, with a
 * little macro magic to make the message optional (defaulting to the
 * stringification of the tested expression).
 *
 * Contrary to BUILD_BUG_ON(), static_assert() can be used at global
 * scope, but requires the expression to be an integer constant
 * expression (i.e., it is not enough that __builtin_constant_p() is
 * true for expr).
 *
 * Also note that BUILD_BUG_ON() fails the build if the condition is
 * true, while static_assert() fails the build if the expression is
 * false.
 */
#define static_assert(expr, ...) __static_assert(expr, ##__VA_ARGS__, #expr)
#define __static_assert(expr, msg, ...) _Static_assert(expr, msg)
```

I hope these info found can help you to do the better solution, good night

**Comment by Fantu** [comment]
@SergeiShtepa I also built kernel for armhf and arm64, always with success with https://github.com/Fantu/linux-blksnap/commit/895a41fcdc62466a79b488e3e75e21f4c44c1869 (even if static assert probably should be improved or changed)
No additional warning (with W=1) but only:
`drivers/block/blksnap/diff_storage.c:30: warning: Function parameter or member 'bdev_path' not described in 'storage_bdev'`
So now I have things ready for fast build on all next patches update on amd64, armhf and arm64 (that I suppose are the most used) and ppc (ppc32 ppc64 ppc64le and also with diffent configs).
Is there other archs you think important to keep tested on any change?

**Comment by Fantu** [comment]
@SergeiShtepa I saw a reply from DM maintainer in v2 patch posted: https://lore.kernel.org/lkml/Y8cNVv4O+vjL+aAy@redhat.com/
And also reply on lwn to the message of Conan_Kudo: https://lwn.net/Articles/914908/ https://lwn.net/Articles/920245/
And other reply to you: https://lwn.net/Articles/915293/
I think it's good to answer in the mailing list of the patch and understand what would be the best way to proceed together with the DM and BLOCK maintainers (and other developers if needed).

**Comment by SergeiShtepa** [comment]
@Fantu  Yep. Thanks. I see. In progress.

**Comment by Fantu** [comment]
@SergeiShtepa hi, I didn't saw replies on https://lore.kernel.org/lkml/20221209142331.26395-1-sergei.shtepa@veeam.com/
are not arrived to ML or you didn't had time to reply for now?

**Comment by SergeiShtepa** [comment]
Hi, @Fantu !
In this case, the most stupid thing that can be done is to write a quick, ill-conceived, emotional response.
We are preparing an answer, don't worry.

**Comment by SergeiShtepa** [comment]
Hi all.

The [upstream version](https://github.com/SergeiShtepa/linux/tree/blksnap-master) of blksnap changes the interface.
A branch [blksnap_interface_v2](https://github.com/veeam/blksnap/tree/blksnap_interface_v2) was created to check the health of the upstream version.
The out-of-tree module in this branch has not been changed yet.

In the future, I would like the user-space code to work equally with both the module in-tree and out-of-tree one.
But at the moment this is a secondary task.

**Comment by SergeiShtepa** [comment]
Hi!
The [patch v3](https://github.com/veeam/blksnap/commit/a7cd0df1d3ab19229381380fc0631100a77debbb) for kernel v6.3.0-rc4 is available for discussion.
I will be glad of your feedback.

**Comment by Fantu** [comment]
@SergeiShtepa good v3 have big changes and and it was also "lightened"

for now I did only a fast look

patch subjects now is better, I think that many of them should have also "add" like:
documentation: **add** Block Device Filtering Mechanism
but maybe I'm wrong and it's not necessary

about MAINTAINERS add of new entries splitted in many commits seems strange to me, it was suggested by Christoph?

MAINTAINERS changes in patch 02/15 still have 2 typo "inlcude" that need to be fixed in "include"

Christoph did many changes, especially blkfilters (but not only) so seems good to me add him with Co-developed-by (see https://docs.kernel.org/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by) at least in blkfilter patch (if you want do it ask to him before)

I hope to have some time in the weekend for do some tests and look better

**Comment by SergeiShtepa** [comment]
> about MAINTAINERS add of new entries splitted in many commits seems strange to me, it was suggested by Christoph?

I do not know how to do it correctly, but it seems to me quite logical. This separation makes patches independent.
A file was added in the commit, information about maintainer was also added in it.
This allowed me to check that all the files were added correctly, which means it will make it easier to check the patch, I think.

> MAINTAINERS changes in patch 02/15 still have 2 typo "inlcude" that need to be fixed in "include"

Thanks, it will be fixed.
> Christoph did many changes, especially blkfilters (but not only) so seems good to me add him with Co-developed-by (see https://docs.kernel.org/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by) at least in blkfilter patch (if you want do it ask to him before)

Yes, I read it. But it remains a mystery to me exactly how to specify co-authors of the project. 
Do I need to use git "magic" or do I need to enter these lines into the commit text?

> I hope to have some time in the weekend for do some tests and look better

That would be great.
I will also take the time to test next week.

**Comment by Fantu** [comment]
> > Christoph did many changes, especially blkfilters (but not only) so seems good to me add him with Co-developed-by (see https://docs.kernel.org/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by) at least in blkfilter patch (if you want do it ask to him before)
> 
> Yes, I read it. But it remains a mystery to me exactly how to specify co-authors of the project. Do I need to use git "magic" or do I need to enter these lines into the commit text?
Seems explained well in kernel doc for me but is possible check also some commits in kernel git, about git tools support add of single Signed-off-by: of the author automatically (if setted), when more are needed should be added manually in commit description

for example if I'm not wrong in this case on "footer" of commit description should be:
```

Co-developed-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Sergei Shtepa <sergei.shtepa@veeam.com>
```
looking many commits seems major of people on patches developed by multiple author use only multiple Signed-off-by: without Co-developed-by: so probably is correct also without Co-developed-by:
anyway should be asked permission to Christoph Hellwig before add and also if he want suggest other fixes before submit

with this big add many changes, many review/suggest, many versions of patch submitted etc... I understand is difficult add all things related  to commit description complete and good but I think is good add atleast Christoph and atleast in blkfilter part for the big contributing gave
and when someone will give its review and/or approvals will also should be added (Reviewed-by: and Acked-by:) but this should be easy as will be specified in the replay, here one example: https://lore.kernel.org/all/e088f7f1-827e-f0e5-4fc0-df0dcefbb873@opensource.wdc.com/

after initial upstream inclusion should be done good in any patch with all needed things including also Reported-by:, Tested-by:, Reviewed-by:, Suggested-by: and Fixes:

About add my "Tested-by:" I think is not good as I only did some build tests and superficial usage tests, as I would also do for this new version, unfortunately it would take a few whole days for complete and serious tests

**Comment by Fantu** [comment]
for now build on amd64 (both builtin and module) and armhf and didn't gave errors or warning

additional kernel-doc comments check with scripts/kernel-doc spotted:
```
drivers/block/blksnap/diff_area.h:106: warning: Function parameter or member 'physical_blksz' not described in 'diff_area'
drivers/block/blksnap/diff_area.h:106: warning: Function parameter or member 'logical_blksz' not described in 'diff_area'
```

**Comment by SergeiShtepa** [comment]
> for now build on amd64 (both builtin and module) and armhf and didn't gave errors or warning
> 
> additional kernel-doc comments check with scripts/kernel-doc spotted:
> 
> ```
> drivers/block/blksnap/diff_area.h:106: warning: Function parameter or member 'physical_blksz' not described in 'diff_area'
> drivers/block/blksnap/diff_area.h:106: warning: Function parameter or member 'logical_blksz' not described in 'diff_area'
> ```
> 
> EDIT: done also build on arm64 and some ppc config

[Thanks!](https://github.com/SergeiShtepa/linux/commit/c388a54c19cd2e449dfbf7dd9f6a137d0ced2b22)

**Comment by Fantu** [comment]
@SergeiShtepa I saw now your latest commits on https://github.com/veeam/blksnap/tree/stable-v2.0, v2 will be only with upstream kernel module and not the external module? or is only temporary thing?

**Comment by SergeiShtepa** [comment]
Hi @Fantu .
I expect that this branch will be the main one, for users, after the appearance of blksnap in the upstream.
I hope this event will happen in the near future. Then it will no longer be necessary to supply blksnap separately for new kernels.
I don't know what I'm going to do in the master branch yet. There is one idea, but I doubt it yet.
I don't plan any other stable branches.

**Comment by SergeiShtepa** [comment]
And @Fantu  - check please my latest commits in branch stable-v2.0.
I have corrected the actions.

**Comment by Fantu** [comment]
@SergeiShtepa from a fast look seems ok

**Comment by SergeiShtepa** [comment]
Hi!
v3 on [patchwork](https://patchwork.kernel.org/project/linux-block/list/?series=737222) and [lore](https://lore.kernel.org/linux-block/20230404140835.25166-1-sergei.shtepa@veeam.com/).
There were some problems with the server 'vger.kernel.org' therefore, the emails appeared on the server only yesterday.

**Comment by Fantu** [comment]
@SergeiShtepa I spotted and write only yesterday about the "Changes:" without the version "in vN" that can be confused with note of previous version above :(
I hope do not cause problems for those who will read.
this version should be better anyway as all the warnings of complete build and other thorough checks have been fixed so no reports of kernel robots should arrive most likely and perhaps it will be better seen by some reviewers who may ignore it if the prerequisites for posting new patches are not met (older versions didn't meet all prerequisites)

**Comment by SergeiShtepa** [comment]
Yes, thanks, I read it.
But pay attention, the date of sending the emails is 04.04.2023.
The process of sending the emails was long this time.

**Comment by Fantu** [comment]
good news, seems that patches for upstream received the first "Tested-by:" of Donald Buczek that seems had done a deep testing

**Comment by SergeiShtepa** [comment]
Hi.
News: v4 patch prepared.
See: https://github.com/veeam/blksnap/tree/master/patches/lk6.4-rc4 
Feedback is welcome!

**Comment by Fantu** [comment]
hi, I haven't continued the usage tests lately because I wanted to disconnect from pc in my free time and in the last few days disconnecting for most of the day for health reasons
I did only a fast build test (amd64 builtin/module and docs) of [lk6.4-rc4-v2](https://github.com/veeam/blksnap/commit/68ee48de00e73ebe1e8006fda29d1f544bc97fd8) and I not found errors and warning from a fast look
@SergeiShtepa have changelog only of latest version in cover I suppose is not a good things, v2 also no longer even has the link to v3

**Comment by SergeiShtepa** [comment]
Hi!
Thanks!
OK, I will make a full log of changes, as in the previous version of the patch.
Be healthy! We all need to be able to listen to our body when it needs help.

**Comment by SergeiShtepa** [comment]
Hi.
patch v4 was sent.
But due to the "unwarmed" graylist on the mail server @vger.kernel.org, the first mail was torn off from the rest of the patchset.
Header: https://lore.kernel.org/linux-block/20230609115206.4649-1-sergei.shtepa@veeam.com/T/#u
Patchset: https://lore.kernel.org/linux-block/20230609115858.4737-1-sergei.shtepa@veeam.com/T/#t
Whether this is a problem or not, I do not know. I will be waiting for comments on this.

**Comment by SergeiShtepa** [comment]
patch v5.
For "kernel/git/axboe/linux-block.git" branch "for-6.5/block". [Link](https://git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git/log/?h=for-6.5/block).
[lore](https://lore.kernel.org/linux-block/ZIcsijGWeyk%2FFjHs@infradead.org/T/#t ).
[patchwork](https://patchwork.kernel.org/project/linux-block/list/?series=756312).

**Comment by Fantu** [comment]
@SergeiShtepa the latest mail shipments seem to have been a bit chaotic^^'' v4 splitted and double v5 I saw and replies "scattered"
there is good news that Christoph Hellwig https://lore.kernel.org/linux-block/ZIcsijGWeyk%2FFjHs@infradead.org/T/#mb373f939657cfe8c2c2439ef4f5886be21bb061c seems consider blksnap ready for inclusion (there are still some things to do from the replies
I hope that mail send issues don't discourage reviewers

**Comment by Fantu** [comment]
looking https://github.com/torvalds/linux after the latest "block merge" of 2 days ago seems that "for-next" of block was fully merged (https://git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git/log/?h=for-next) so I think can blksnap can be rebased on latest https://github.com/torvalds/linux (now that include all the block changes) for working on next version of the patch

**Comment by SergeiShtepa** [comment]
Yes, thanks Fabio.
We already have 6.5-rc2 .
I apologize for the late response. I was on vacation.
There's still a lot of work to do on redesigning the ioctl extension of the difference storage. And blk-crypto...

**Comment by SergeiShtepa** [comment]
Hi!
Little update.

In the ['stable-v2.0'](https://github.com/veeam/blksnap/tree/stable-v2.0) branch of this repository and in the branch ['blksnap-master'](https://github.com/SergeiShtepa/linux/tree/blksnap-master) based on vanilla 6.6-rc2 of [linux](https://github.com/SergeiShtepa/linux/) repository , the concept of storing snapshot differences in a file was implemented. 
The ability to create a difference storage on a block device under the snapshot has been removed.
Storing differences on a block device remains possible, but only in the case of exclusive ownership of it.

An interesting consequence of these changes:
* Now it is possible to create a difference storage file with the O_TMPFILE flag. The file will be automatically deleted when its descriptor is closed when the snapshot is released.
* Now it is possible to use files in memory and in the 'swap' partition thanks to files on tmpfs.

I think that it is possible to return the ability to create a difference storage on a block device under snapshots. But this will require support from file systems. It can be quite a difficult job.
I plan to start it when I realize that the blksnap module is in demand by users and they really need this feature.

In the near future, I will be engaged in testing, documentation improvement and preparing a patch.
I invite everyone to take part in testing the current version.

**Comment by SergeiShtepa** [comment]
Hi.
Patch [lk6.7-rc1-v2](https://github.com/veeam/blksnap/tree/master/patches/lk6.7-rc1-v2) has been prepared.
I plan to send it next week.

**Comment by Fantu** [comment]
Tested build on amd64 (both builtin and module), armhf, arm64 and various ppc config.
I don't have time to deep check and testing it in these days but seems to me achieved the main requirements for integration, and that with the latest changes reviewer had recommended I assume is approaching.
@SergeiShtepa before send, I suggest taking a look at https://github.com/veeam/blksnap/pull/79 relating to the patches cover

**Comment by SergeiShtepa** [comment]
V6:
[lore.kernel.org/](https://lore.kernel.org/linux-block/14d5d31e-0dbe-8d04-91a6-82a886f8e92a@veeam.com/T/#t)
[patchwork.kernel.org](https://patchwork.kernel.org/project/linux-block/list/?series=804089)

**Comment by SergeiShtepa** [comment]
Hi.
Patch [lk6.8-rc1](https://github.com/veeam/blksnap/tree/master/patches/lk6.8-rc1) has been prepared.
It also available as a branch in my [fork](https://github.com/SergeiShtepa/linux/tree/blksnap-lk6.8-rc1).

**Comment by Fantu** [comment]
Hi, I don't have enough time to test its use, I just launched some amd64 builds (builtin and module), armhf, arm64 and some ppc. (from automatic tests I prepared in past).
One ppc build failed for an issue not related to blksnap.
On armhf build (https://github.com/Fantu/linux-blksnap/actions/runs/7629304370/job/20782415299) I spotted this warning:
```
2024-01-23T17:55:07.5898286Z drivers/block/blksnap/main.c: In function â€˜ioctl_snapshot_createâ€™:
2024-01-23T17:55:07.5901218Z ##[warning]drivers/block/blksnap/main.c:196:30: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
2024-01-23T17:55:07.5903698Z   196 |         fname = strndup_user((const char __user *)karg.diff_storage_filename,
2024-01-23T17:55:07.5904914Z       |                              ^
```

**Comment by SergeiShtepa** [comment]
v7
https://patchwork.kernel.org/project/linux-block/list/?series=824711
https://lore.kernel.org/all/20240209160204.1471421-1-sergei.shtepa@linux.dev/

**Comment by Fantu** [comment]
This time the send of patch serie seems complete and without duplicate 👍

**Comment by SergeiShtepa** [comment]
Yep.
I have reduce the number of patches in set and greatly reduced the list of recipients.
The script get_maintainer.pl returns such a long list that it can no longer be used to generate a list of recipients.
The test bot should respond soon. He will not be able to apply patches to the master branch and build.

**Comment by Fantu** [comment]
Looking kernel bot message of older version seems it try also with linux-block for-next: https://lore.kernel.org/lkml/202211031239.9fxTTo7W-lkp@intel.com/

> url:    https://github.com/intel-lab-lkp/linux/commits/Sergei-Shtepa/blksnap-creating-non-persistent-snapshots-for-backup/20221103-004434
> base:   https://git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git for-next

So I think that if will apply correctly with master or linux-block don't report error, unless it detect other errors and/or warnings

**Comment by SergeiShtepa** [comment]
I checked it out.
The patch can be successfully applied to Jens branch block/for-next and can be successfully build.
Jens create this branch at the end of last week.

**Comment by Fantu** [comment]
Of v7 I didn't pay attention to the patches subjects, from patch 3/8 to patch 8/8 I think it should be "blksnap:" or "block/blksnap:" instead of "block:" at the beginning of the subjects

**Comment by Fantu** [comment]
@SergeiShtepa hi, Christoph did write to you to find out if the latest version was good enough for the upstream application, or need other improvements, or did he not have time to review it?
I haven't seen any public responses to the latest version, I suspect that someone who was interested missed it due to some patches with incorrect titles.
Perhaps even with the fewest mail recipients, but that seems necessary to avoid it being blocked and being sent back double (or more), which would annoy the recipients :(

**Comment by SergeiShtepa** [comment]
Hi, @Fantu !

> I haven't seen any public responses to the latest version, I suspect that someone who was interested missed it due to some patches with incorrect titles.

Yep. Indeed, there is no reaction. I think that everyone who wanted to see the latest patchset did it, regardless of whether he is the recipient of the letter or not.

I am currently preparing an updated standalone version of the module for one of the next releases of our product.
I want to move a some ideas from the upstream version to it. This will improve performance.

Even if the blksnap module is never accepted into the kernel, the work has not been done in vain.
The code of the blksnap module only gets better from this.

I will definitely return to work on the upstream version later. 
In working on the next patch for upstream, I think I will focus on more detailed testing and performance measurement.

**Comment by SergeiShtepa** [comment]
Hi!
The [patchset](https://github.com/veeam/blksnap/tree/master/patches/lk6.9) for LK v6.9 is ready to view.
The blksnap-lk6.9 [branch](https://github.com/SergeiShtepa/linux/tree/blksnap-lk6.9) is available too.

**Comment by Fantu** [comment]
Hi, this patchset is only for testing on 6.9 but not for submit, right?
A v8 will be prepared for 6.10 to submit?

**Comment by fsupper** [comment]
Hi, any update on this Topic?
What is missing to get the blksnap module into the linux kernel?
Regards
Flo

**Comment by SergeiShtepa** [comment]
Hi. 
> What is missing to get the blksnap module into the linux kernel?

And that's a good question.

As usual, the most valuable resource is missing - time.
But this is a matter of priorities. Well, the most important of them is: "Why?".
I know what Veeam needs it for, but it's completely unclear to me if the upstream needs it.
I have formed the opinion that this is not interesting to the Linux community.
Other backup vendors are not involved in the patchset discussion.
Everyone seems to be quite satisfied with dm-snap.

Nevertheless, I plan to make at least one more attempt. It is possible that this will happen this autumn.

**Comment by fsupper** [comment]
> Hi.
> 
> > What is missing to get the blksnap module into the linux kernel?
> 
> And that's a good question.
> 
> As usual, the most valuable resource is missing - time. But this is a matter of priorities. Well, the most important of them is: "Why?". I know what Veeam needs it for, but it's completely unclear to me if the upstream needs it. I have formed the opinion that this is not interesting to the Linux community. Other backup vendors are not involved in the patchset discussion. Everyone seems to be quite satisfied with dm-snap.
> 
> Nevertheless, I plan to make at least one more attempt. It is possible that this will happen this autumn.

Thanks for your honest answer!
regards

**Comment by SergeiShtepa** [comment]
Yes, that's right. I agree with every word @Fantu.

Plus, I and the company I work for have pretty little experience working with open source.
Perhaps , this is our point of growth. I suppose that more efforts should be made in this direction.
This is just my personal opinion.

**Comment by SergeiShtepa** [comment]
Hi.
A patch has been prepared for the 6.15 kernel.
https://github.com/veeam/blksnap/tree/master/patches/lk6.15

The patch is currently under testing.
More information about the changes can be found in the history here.
https://github.com/SergeiShtepa/linux/tree/blksnap-lk6.15

**Comment by Fantu** [comment]
Big thanks for your work.
Before send the patches can you post the serie including the header? So will possible to detect things to fix/improve and try to avoid for example some not good title of previous sent.

**Comment by SergeiShtepa** [comment]
> Big thanks for your work. Before send the patches can you post the serie including the header? So will possible to detect things to fix/improve and try to avoid for example some not good title of previous sent.

Yes, of course. Thank you.

**Comment by Fantu** [comment]
I did some fast automatic checks, with codespell no typo spotted, about build on the armhf one I found this warning:
```
2025-05-07T22:16:07.4639988Z drivers/block/blksnap/main.c: In function ‘ioctl_snapshot_create’:
2025-05-07T22:16:07.4642265Z ##[warning]drivers/block/blksnap/main.c:188:30: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
2025-05-07T22:16:07.4644047Z   188 |         fname = strndup_user((const char __user *)karg.diff_storage_filename,
2025-05-07T22:16:07.4644672Z       |       
```

**Comment by SergeiShtepa** [comment]
Thanks.

armhf - 32 bit arch.
This fix should help:
https://github.com/veeam/blksnap/blob/master/patches/lk6.15/0003-use-u64_to_user_ptr-instead-cast-to-pointer-from-int.patch

**Comment by Fantu** [comment]
Another check that spotted something:
```
./scripts/kernel-doc -v -none drivers/block/blksnap/*
...
drivers/block/blksnap/cbt_map.h:78: warning: Function parameter or struct member 'bdev_capacity' not described in 'cbt_map'
drivers/block/blksnap/cbt_map.h:78: warning: Excess struct member 'device_capacity' description in 'cbt_map'
drivers/block/blksnap/diff_area.h:124: warning: Function parameter or struct member 'image_io_queue_count' not described in 'diff_area'
```

**Comment by SergeiShtepa** [comment]
> Another check that spotted something:

Thank you.
I will definitely fix this.
I've just had a lot of work lately...

**Comment by Fantu** [comment]
I tried to post upstream to know whether there are reasons why the latest version has not been revised and whether there might be reasons to oppose it:
https://lore.kernel.org/linux-block/5459ee8c-075f-453f-8999-19219f9a6fdc@tiscali.it/T/#u

From replies, Christoph Hellwig thinks it would be useful and seems still interested, and Jens Axboe says he wasn't personally convinced but is open to taking another look if someone rebases and reposts the patchset. His main concern is the "dump-and-run" pattern, that once the company goal of getting it upstream is achieved, motivation to maintain it disappears.

@SergeiShtepa What do you think?

**Comment by SergeiShtepa** [comment]
Thank you.  
Yes, I have read the emails.  

What do I think… Hmm… This is a very complicated question.  

**I’ll share some of my personal thoughts. These do not represent the position of the company I work for. Moreover, some of these thoughts are just my assumptions and may be incorrect.**

First of all, I think I’m not passionate enough about the Linux kernel. There are plenty of other interesting tasks at my work, and in my free time, I try to rest.  To be a maintainer of some part of the kernel, you really need to dedicate a lot more time to the project.  
I believe that Jens doesn’t need so much new code as he needs a colleague he can rely on someone who can find bugs in other people’s code, do reviews, and help him. He doesn’t see such a person in me.

Secondly, blksnap doesn’t solve the main task that Jens is focused on. If you look at his work, it’s primarily about optimizing the block layer and speeding up I/O operations. His utility, fio, has become a standard for performance evaluation. He doesn’t need an extra handler in the block layer.

When I proposed the kernel module, I was hoping for support from others who, as I thought, were interested parties.  
However, some people are convinced that Device Mapper already solves the snapshot problem quite well. The maintainers of this part of the kernel don’t see a need to expand its functionality. Others believe that snapshots should be handled at the file system level, for example with BTRFS, and suggest solutions based on that technology. As for the companies developing backup solutions, I can only say they are too busy competing with each other and don’t want to join forces to develop a common system component. And that’s really difficult.

Nevertheless, the work on the blksnap module proposal has brought significant benefits to the project. The code has become much better. Compare it to veeamsnap and you’ll see a huge difference. In version 13, optimizations suggested by Christoph were implemented, which significantly improved performance. This is already bringing great benefits to people.

The module is becoming more in demand. To my knowledge, it’s used not only in the Veeam project. Maybe in the future, when it becomes more popular and more widely used, we’ll propose it for upstream again.

**Comment by Fantu** [comment]
Thanks for sharing your thoughts Sergei.

I think there might be a small misunderstanding about Jens' concern though.
He's not asking you to become a block layer maintainer.
His worry is specifically about the blksnap code itself: if it gets merged, will someone keep maintaining it, or will it be abandoned after the merge? He wants to know there's a committed maintainer for the blksnap part, not for the whole block subsystem.
Unfortunately, there have already been cases where companies got their code upstream and then abandoned or neglected the maintenance of the parts they had added (a recent notable example is Intel, whose layoffs left several kernel drivers orphaned, including coretemp and ISST). Greg Kroah-Hartman regularly has to remove tens of thousands of lines of unmaintained code from the staging tree. So Jens' concern about "dump-and-run" is based on real experience, not a hypothetical fear.

**Comment by Foosec** [comment]
I check up on this every year or so, its something that is sorely missing from Linux, especially from a desktop backup perspective. Hopefully things move forward eventually
